File: frontend/lib/services/payment_service.dart
import 'api_client.dart';

// ── Models ────────────────────────────────────────────────────────────────────

class PlanModel {
  final String id;
  final String adminId;
  final String name;
  final String? description;
  final double price;
  final String currency;
  final int durationMonths;
  final String? modality;
  final int evaluationsIncluded;
  final bool isActive;
  final DateTime createdAt;

  PlanModel({
    required this.id,
    required this.adminId,
    required this.name,
    this.description,
    required this.price,
    required this.currency,
    required this.durationMonths,
    this.modality,
    required this.evaluationsIncluded,
    required this.isActive,
    required this.createdAt,
  });

  factory PlanModel.fromJson(Map<String, dynamic> json) {
    return PlanModel(
      id: json['id'] as String,
      adminId: json['admin_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      price: double.parse(json['price'].toString()),
      currency: json['currency'] as String? ?? 'BRL',
      durationMonths: json['duration_months'] as int,
      modality: json['modality'] as String?,
      evaluationsIncluded: json['evaluations_included'] as int? ?? 0,
      isActive: json['is_active'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  String get durationLabel {
    switch (durationMonths) {
      case 1: return '1 mês';
      case 3: return '3 meses';
      case 6: return '6 meses';
      case 12: return '1 ano';
      default: return '$durationMonths meses';
    }
  }

  String get priceFormatted => 'R\$ ${price.toStringAsFixed(2).replaceAll('.', ',')}';
}

class SubscriptionModel {
  final String id;
  final String studentId;
  final String planId;
  final String adminId;
  final String status;
  final String? paymentMethod;
  final DateTime? startedAt;
  final DateTime? expiresAt;
  final DateTime createdAt;
  final PlanModel? plan;

  SubscriptionModel({
    required this.id,
    required this.studentId,
    required this.planId,
    required this.adminId,
    required this.status,
    this.paymentMethod,
    this.startedAt,
    this.expiresAt,
    required this.createdAt,
    this.plan,
  });

  factory SubscriptionModel.fromJson(Map<String, dynamic> json) {
    return SubscriptionModel(
      id: json['id'] as String,
      studentId: json['student_id'] as String,
      planId: json['plan_id'] as String,
      adminId: json['admin_id'] as String,
      status: json['status'] as String,
      paymentMethod: json['payment_method'] as String?,
      startedAt: json['started_at'] != null ? DateTime.parse(json['started_at'] as String) : null,
      expiresAt: json['expires_at'] != null ? DateTime.parse(json['expires_at'] as String) : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      plan: json['plan'] != null ? PlanModel.fromJson(json['plan'] as Map<String, dynamic>) : null,
    );
  }

  bool get isActive => status == 'active';
  bool get isPending => status == 'pending';
  bool get isCanceledPending => status == 'canceled_pending';
  bool get isExpired => status == 'expired';
  bool get isCanceled => status == 'canceled';
  bool get canRenew => isExpired || isCanceled;

  String get statusLabel {
    switch (status) {
      case 'active': return 'Ativa';
      case 'pending': return 'Pendente';
      case 'expired': return 'Expirada';
      case 'canceled_pending': return 'Cancelando';
      case 'canceled': return 'Cancelada';
      default: return status;
    }
  }

  int get daysRemaining {
    if (expiresAt == null) return 0;
    return expiresAt!.difference(DateTime.now()).inDays;
  }
}

class SubscriptionSummary {
  final int totalActive;
  final int totalPending;
  final int totalCanceled;
  final int totalExpired;
  final double revenueThisMonth;
  final double revenueLastMonth;

  SubscriptionSummary({
    required this.totalActive,
    required this.totalPending,
    required this.totalCanceled,
    required this.totalExpired,
    required this.revenueThisMonth,
    required this.revenueLastMonth,
  });

  factory SubscriptionSummary.fromJson(Map<String, dynamic> json) {
    return SubscriptionSummary(
      totalActive: json['total_active'] as int? ?? 0,
      totalPending: json['total_pending'] as int? ?? 0,
      totalCanceled: json['total_canceled'] as int? ?? 0,
      totalExpired: json['total_expired'] as int? ?? 0,
      revenueThisMonth: double.parse(json['revenue_this_month'].toString()),
      revenueLastMonth: double.parse(json['revenue_last_month'].toString()),
    );
  }
}

class AdminSubscriptionItem {
  final String id;
  final String studentId;
  final String studentName;
  final String studentEmail;
  final String planName;
  final double planPrice;
  final String status;
  final String? paymentMethod;
  final DateTime? startedAt;
  final DateTime? expiresAt;
  final DateTime createdAt;

  AdminSubscriptionItem({
    required this.id,
    required this.studentId,
    required this.studentName,
    required this.studentEmail,
    required this.planName,
    required this.planPrice,
    required this.status,
    this.paymentMethod,
    this.startedAt,
    this.expiresAt,
    required this.createdAt,
  });

  factory AdminSubscriptionItem.fromJson(Map<String, dynamic> json) {
    return AdminSubscriptionItem(
      id: json['id'] as String,
      studentId: json['student_id'] as String,
      studentName: json['student_name'] as String,
      studentEmail: json['student_email'] as String,
      planName: json['plan_name'] as String,
      planPrice: double.parse(json['plan_price'].toString()),
      status: json['status'] as String,
      paymentMethod: json['payment_method'] as String?,
      startedAt: json['started_at'] != null ? DateTime.parse(json['started_at'] as String) : null,
      expiresAt: json['expires_at'] != null ? DateTime.parse(json['expires_at'] as String) : null,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  String get statusLabel {
    switch (status) {
      case 'active': return 'Ativa';
      case 'pending': return 'Pendente';
      case 'expired': return 'Expirada';
      case 'canceled_pending': return 'Cancelando';
      case 'canceled': return 'Cancelada';
      default: return status;
    }
  }

  bool get isActive => status == 'active';
  bool get isPending => status == 'pending';
}

class CheckoutResponse {
  final String subscriptionId;
  final String checkoutUrl;
  final String externalPaymentId;
  final String status;

  CheckoutResponse({
    required this.subscriptionId,
    required this.checkoutUrl,
    required this.externalPaymentId,
    required this.status,
  });

  factory CheckoutResponse.fromJson(Map<String, dynamic> json) {
    return CheckoutResponse(
      subscriptionId: json['subscription_id'] as String,
      checkoutUrl: json['checkout_url'] as String,
      externalPaymentId: json['external_payment_id'] as String,
      status: json['status'] as String,
    );
  }
}

// ── Service ───────────────────────────────────────────────────────────────────

class PaymentService {
  final ApiClient _apiClient;

  PaymentService({required ApiClient apiClient}) : _apiClient = apiClient;

  Future<List<PlanModel>> getAvailablePlans() async {
    final response = await _apiClient.get<List<PlanModel>>(
      '/plans',
      fromJson: (data) {
        if (data is List) {
          return data
              .whereType<Map<String, dynamic>>()
              .map((e) => PlanModel.fromJson(e))
              .toList();
        }
        return [];
      },
    );
    return response;
  }

  Future<List<PlanModel>> getAdminPlans() async {
    final response = await _apiClient.get<List<PlanModel>>(
      '/admin/plans',
      fromJson: (data) {
        if (data is List) {
          return data
              .whereType<Map<String, dynamic>>()
              .map((e) => PlanModel.fromJson(e))
              .toList();
        }
        return [];
      },
    );
    return response;
  }

  Future<PlanModel> createPlan({
    required String name,
    String? description,
    required double price,
    required int durationMonths,
    String? modality,
    int evaluationsIncluded = 0,
  }) async {
    return await _apiClient.post<PlanModel>(
      '/admin/plans',
      body: {
        'name': name,
        'description': ?description,
        'price': price,
        'duration_months': durationMonths,
        'modality': ?modality,
        'evaluations_included': evaluationsIncluded,
      },
      fromJson: (data) => PlanModel.fromJson(data as Map<String, dynamic>),
    );
  }

  Future<PlanModel> updatePlan(String planId, {
    String? name,
    String? description,
    double? price,
    int? durationMonths,
    bool? isActive,
  }) async {
    return await _apiClient.put<PlanModel>(
      '/admin/plans/$planId',
      body: {
        'name': ?name,
        'description': ?description,
        'price': ?price,
        'duration_months': ?durationMonths,
        'is_active': ?isActive,
      },
      fromJson: (data) => PlanModel.fromJson(data as Map<String, dynamic>),
    );
  }

  Future<void> deletePlan(String planId) async {
    await _apiClient.delete<void>(
      '/admin/plans/$planId',
      fromJson: (_) {},
    );
  }

  Future<CheckoutResponse> createCheckout({
    required String planId,
    required String paymentMethod,
    String? replacementPolicy,
  }) async {
    return await _apiClient.post<CheckoutResponse>(
      '/subscriptions/checkout',
      body: {
        'plan_id': planId,
        'payment_method': paymentMethod,
        'replacement_policy': ?replacementPolicy,
      },
      fromJson: (data) => CheckoutResponse.fromJson(data as Map<String, dynamic>),
    );
  }

  Future<SubscriptionModel?> getCurrentSubscription() async {
    try {
      return await _apiClient.get<SubscriptionModel>(
        '/subscriptions/current',
        fromJson: (data) => SubscriptionModel.fromJson(data as Map<String, dynamic>),
      );
    } on NotFoundException {
      return null;
    } catch (e) {
      return null;
    }
  }

  Future<bool> hasActiveSubscription() async {
    final sub = await getCurrentSubscription();
    return sub?.isActive ?? false;
  }

  Future<bool> cancelMySubscription() async {
    try {
      await _apiClient.post<Map<String, dynamic>>(
        '/subscriptions/cancel',
        body: {},
        fromJson: (data) => data as Map<String, dynamic>,
      );
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<SubscriptionSummary> getDashboardSummary() async {
    return await _apiClient.get<SubscriptionSummary>(
      '/admin/subscriptions/summary',
      fromJson: (data) => SubscriptionSummary.fromJson(data as Map<String, dynamic>),
    );
  }

  Future<List<AdminSubscriptionItem>> getDashboardSubscriptions({
    String? statusFilter,
    int limit = 50,
    int offset = 0,
  }) async {
    final params = <String, dynamic>{
      'limit': limit,
      'offset': offset,
      'status_filter': ?statusFilter,
    };
    return await _apiClient.get<List<AdminSubscriptionItem>>(
      '/admin/subscriptions/dashboard',
      queryParameters: params,
      fromJson: (data) {
        if (data is List) {
          return data
              .whereType<Map<String, dynamic>>()
              .map((e) => AdminSubscriptionItem.fromJson(e))
              .toList();
        }
        return [];
      },
    );
  }

  Future<bool> manualActivate(String subscriptionId) async {
    try {
      await _apiClient.post<Map>(
        '/admin/subscriptions/$subscriptionId/activate',
        body: {},
        fromJson: (data) => data as Map,
      );
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<bool> manualCancel(String subscriptionId) async {
    try {
      await _apiClient.post<Map>(
        '/admin/subscriptions/$subscriptionId/cancel',
        body: {},
        fromJson: (data) => data as Map,
      );
      return true;
    } catch (_) {
      return false;
    }
  }

  Future<bool> changePlan(String subscriptionId, String newPlanId) async {
    try {
      await _apiClient.put<Map>(
        '/admin/subscriptions/$subscriptionId/change-plan',
        body: {'plan_id': newPlanId},
        fromJson: (data) => data as Map,
      );
      return true;
    } catch (_) {
      return false;
    }
  }
}

File: backend/app/services/payment_service.py
"""
Service layer para Pagamentos (MVP V1)
Contém lógica de negócio e orquestração
"""
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional, List
from uuid import UUID
from decimal import Decimal
from datetime import datetime, timedelta
import logging

from app.models.payment import Plan, Subscription
from app.repositories.payment_repository import PlanRepository, SubscriptionRepository
from app.dtos.payment_dtos import (
    CreatePlanDTO,
    UpdatePlanDTO,
    PlanResponseDTO,
    SubscriptionResponseDTO,
    SubscriptionDetailDTO,
    CheckoutResponseDTO,
    SubscriptionSummaryDTO,
    AdminSubscriptionItemDTO
)

logger = logging.getLogger(__name__)


class PlanService:
    """Serviço de gerenciamento de planos"""

    @staticmethod
    async def create_plan(
        session: AsyncSession,
        admin_id: UUID,
        dto: CreatePlanDTO
    ) -> PlanResponseDTO:
        """Criar novo plano"""
        plan = await PlanRepository.create(session, admin_id, dto)
        await session.commit()
        logger.info(f"Plano criado: {plan.id} por admin {admin_id}")
        return PlanResponseDTO.model_validate(plan)

    @staticmethod
    async def get_plan(session: AsyncSession, plan_id: UUID) -> Optional[PlanResponseDTO]:
        """Buscar plano por ID"""
        plan = await PlanRepository.find_by_id(session, plan_id)
        if not plan:
            logger.warning(f"Plano não encontrado: {plan_id}")
            return None
        return PlanResponseDTO.model_validate(plan)

    @staticmethod
    async def list_plans(
        session: AsyncSession,
        admin_id: UUID,
        only_active: bool = True
    ) -> List[PlanResponseDTO]:
        """Listar planos de um admin"""
        plans = await PlanRepository.find_by_admin(session, admin_id, only_active)
        return [PlanResponseDTO.model_validate(p) for p in plans]

    @staticmethod
    async def update_plan(
        session: AsyncSession,
        plan_id: UUID,
        dto: UpdatePlanDTO
    ) -> Optional[PlanResponseDTO]:
        """Atualizar plano"""
        plan = await PlanRepository.update(session, plan_id, dto)
        if not plan:
            logger.warning(f"Plano não encontrado para atualizar: {plan_id}")
            return None
        await session.commit()
        logger.info(f"Plano atualizado: {plan_id}")
        return PlanResponseDTO.model_validate(plan)

    @staticmethod
    async def list_plans_for_student(
        session: AsyncSession,
        student_id: UUID
    ) -> List[PlanResponseDTO]:
        """
        Listar planos ativos para o aluno.
        Cadeia: aluno → trainer_id → invitation.trainer_id → admin que criou o convite
        Se não encontrar admin via convite, retorna todos os planos ativos do sistema.
        """
        from sqlalchemy import select
        from app.models.user import User
        from app.models.invitation import Invitation

        # 1. Buscar o aluno e seu trainer
        result = await session.execute(select(User).where(User.id == student_id))
        student = result.scalar_one_or_none()
        if not student or not student.trainer_id:
            # Sem trainer vinculado — retorna todos os planos ativos
            plans = await PlanRepository.find_all_active(session)
            return [PlanResponseDTO.model_validate(p) for p in plans]

        # 2. Buscar quem convidou o trainer (esse é o admin)
        inv_result = await session.execute(
            select(Invitation).where(
                Invitation.used_by_id == student.trainer_id,
                Invitation.used == True
            )
        )
        invitation = inv_result.scalar_one_or_none()

        if invitation:
            # O admin é quem criou o convite
            admin_result = await session.execute(
                select(User).where(User.id == invitation.trainer_id)
            )
            admin = admin_result.scalar_one_or_none()
            if admin and admin.role == "admin":
                plans = await PlanRepository.find_by_admin(session, admin.id, only_active=True)
                return [PlanResponseDTO.model_validate(p) for p in plans]

        # Fallback: planos do próprio trainer (caso ele seja admin autônomo)
        plans = await PlanRepository.find_by_admin(session, student.trainer_id, only_active=True)
        if plans:
            return [PlanResponseDTO.model_validate(p) for p in plans]

        # Último fallback: todos os planos ativos
        plans = await PlanRepository.find_all_active(session)
        return [PlanResponseDTO.model_validate(p) for p in plans]

    @staticmethod
    async def delete_plan(session: AsyncSession, plan_id: UUID) -> bool:
        """Deletar plano (soft delete)"""
        result = await PlanRepository.soft_delete(session, plan_id)
        if result:
            await session.commit()
            logger.info(f"Plano deletado (soft): {plan_id}")
        return result


class SubscriptionService:
    """Serviço de gerenciamento de assinaturas"""

    @staticmethod
    async def create_checkout(
        session: AsyncSession,
        student_id: UUID,
        plan_id: UUID,
        payment_method: str,
        replacement_policy: Optional[str] = None
    ) -> Optional[CheckoutResponseDTO]:
        """
        Criar checkout para assinatura via InfinitePay.
        Cria o link de pagamento e envia pelo WhatsApp do aluno.
        """
        from sqlalchemy import select
        from app.models.user import User
        from app.services.infinitepay_service import create_payment_link
        from app.services.whatsapp_service import send_payment_link

        # 1. Validar plano
        plan = await PlanRepository.find_by_id(session, plan_id)
        if not plan or not plan.is_active:
            logger.warning(f"Plano inválido: {plan_id}")
            return None

        # 2. Buscar dados do aluno
        result = await session.execute(select(User).where(User.id == student_id))
        student = result.scalar_one_or_none()
        if not student:
            logger.warning(f"Aluno não encontrado: {student_id}")
            return None

        # 3. Criar subscription em status PENDING
        subscription = await SubscriptionRepository.create(
            session,
            student_id,
            plan_id,
            plan.admin_id,
            payment_method,
            replacement_policy
        )
        await session.commit()
        logger.info(f"Assinatura criada (PENDING): {subscription.id}")

        # 4. Criar link de pagamento na InfinitePay
        infinitepay_result = await create_payment_link(
            subscription_id=subscription.id,
            plan_name=plan.name,
            price_brl=float(plan.price),
            student_name=student.name or student.email,
            student_email=student.email,
            student_phone=getattr(student, "phone_whatsapp", None),
        )

        if infinitepay_result:
            checkout_url = infinitepay_result.url
            external_id = f"infinitepay_{subscription.id}"

            # 5. Enviar link via WhatsApp se aluno tiver telefone
            phone = getattr(student, "phone_whatsapp", None) or getattr(student, "phone", None)
            if phone:
                await send_payment_link(
                    phone=phone,
                    student_name=student.name or student.email,
                    plan_name=plan.name,
                    payment_url=checkout_url,
                    price_brl=float(plan.price),
                )
        else:
            # InfinitePay falhou — retorna URL de fallback para o aluno copiar
            logger.warning(f"InfinitePay falhou para subscription {subscription.id}, usando fallback")
            checkout_url = f"https://checkout.infinitepay.io/natalia-faria-16"
            external_id = f"manual_{subscription.id}"

        return CheckoutResponseDTO(
            subscription_id=subscription.id,
            checkout_url=checkout_url,
            external_payment_id=external_id,
            status="pending"
        )

    @staticmethod
    async def get_subscription(
        session: AsyncSession,
        subscription_id: UUID
    ) -> Optional[SubscriptionDetailDTO]:
        """Buscar assinatura com plano"""
        subscription = await SubscriptionRepository.find_by_id(session, subscription_id)
        if not subscription:
            logger.warning(f"Assinatura não encontrada: {subscription_id}")
            return None

        # Carregar plano
        await session.refresh(subscription, ["plan"])
        plan_dto = PlanResponseDTO.model_validate(subscription.plan) if subscription.plan else None

        return SubscriptionDetailDTO(
            **SubscriptionResponseDTO.model_validate(subscription).model_dump(),
            plan=plan_dto
        )

    @staticmethod
    async def get_student_active_subscription(
        session: AsyncSession,
        student_id: UUID
    ) -> Optional[SubscriptionDetailDTO]:
        """Buscar assinatura mais recente do aluno (ativa, expirada ou cancelada)"""
        subscription = await SubscriptionRepository.find_student_latest(session, student_id)
        if not subscription:
            return None

        return await SubscriptionService.get_subscription(session, subscription.id)

    @staticmethod
    async def list_subscriptions(
        session: AsyncSession,
        admin_id: UUID,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0
    ) -> List[SubscriptionDetailDTO]:
        """Listar assinaturas de um admin"""
        subscriptions = await SubscriptionRepository.find_by_admin(
            session,
            admin_id,
            status,
            limit,
            offset
        )

        result = []
        for sub in subscriptions:
            await session.refresh(sub, ["plan"])
            plan_dto = PlanResponseDTO.model_validate(sub.plan) if sub.plan else None
            detail_dto = SubscriptionDetailDTO(
                **SubscriptionResponseDTO.model_validate(sub).model_dump(),
                plan=plan_dto
            )
            result.append(detail_dto)

        return result

    @staticmethod
    async def activate_subscription(
        session: AsyncSession,
        subscription_id: UUID,
        external_payment_id: str
    ) -> Optional[SubscriptionDetailDTO]:
        """
        Ativar assinatura após webhook de pagamento confirmado
        Chamado quando Asaas confirma o pagamento
        """
        subscription = await SubscriptionRepository.find_by_id(session, subscription_id)
        if not subscription:
            logger.warning(f"Assinatura não encontrada para ativar: {subscription_id}")
            return None

        if subscription.status != "pending":
            logger.warning(f"Assinatura não está em PENDING: {subscription_id}")
            return None

        # Carregar plano
        await session.refresh(subscription, ["plan"])
        plan = subscription.plan

        # Ativar
        success = await SubscriptionRepository.activate(
            session,
            subscription_id,
            external_payment_id,
            plan
        )
        
        if success:
            # Se a política for 'immediate', desativamos as outras agora
            if subscription.replacement_policy == "immediate":
                from sqlalchemy import update, and_
                stmt = (
                    update(Subscription)
                    .where(
                        and_(
                            Subscription.student_id == subscription.student_id,
                            Subscription.id != subscription.id,
                            Subscription.status == "active"
                        )
                    )
                    .values(status="canceled", canceled_at=datetime.utcnow())
                )
                await session.execute(stmt)
            
            await session.commit()
            logger.info(f"Assinatura ativada: {subscription_id}")
            return await SubscriptionService.get_subscription(session, subscription_id)

        return None

    @staticmethod
    async def cancel_subscription(
        session: AsyncSession,
        subscription_id: UUID
    ) -> bool:
        """Cancelar assinatura (status CANCELED_PENDING)"""
        result = await SubscriptionRepository.cancel_pending(session, subscription_id)
        if result:
            await session.commit()
            logger.info(f"Assinatura cancelada (PENDING): {subscription_id}")
        return result

    @staticmethod
    async def check_student_access(
        session: AsyncSession,
        student_id: UUID
    ) -> bool:
        """
        Verificar se aluno tem acesso ao app
        Retorna True se tem assinatura ACTIVE e não expirada
        """
        subscription = await SubscriptionRepository.find_student_active(session, student_id)
        if not subscription:
            return False
        return subscription.is_active_now()


    @staticmethod
    async def get_dashboard_summary(
        session: AsyncSession,
        admin_id: UUID
    ) -> SubscriptionSummaryDTO:
        """Resumo financeiro para o dashboard do admin"""
        return await SubscriptionRepository.get_summary(session, admin_id)

    @staticmethod
    async def list_subscriptions_dashboard(
        session: AsyncSession,
        admin_id: UUID,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0
    ) -> List[AdminSubscriptionItemDTO]:
        """Lista de assinaturas com dados do aluno para o dashboard"""
        return await SubscriptionRepository.find_with_student_info(
            session, admin_id, status, limit, offset
        )

    @staticmethod
    async def manual_activate(
        session: AsyncSession,
        subscription_id: UUID
    ) -> bool:
        """Ativar assinatura manualmente (admin sem webhook)"""
        subscription = await SubscriptionRepository.find_by_id(session, subscription_id)
        if not subscription:
            return False
        await session.refresh(subscription, ["plan"])
        success = await SubscriptionRepository.manual_activate(
            session, subscription_id, subscription.plan
        )
        if success:
            await session.commit()
            logger.info(f"Assinatura ativada manualmente: {subscription_id}")
        return success

    @staticmethod
    async def manual_cancel(
        session: AsyncSession,
        subscription_id: UUID
    ) -> bool:
        """Cancelar assinatura manualmente (admin)"""
        success = await SubscriptionRepository.manual_cancel(session, subscription_id)
        if success:
            await session.commit()
            logger.info(f"Assinatura cancelada manualmente: {subscription_id}")
        return success

    @staticmethod
    async def change_plan(
        session: AsyncSession,
        subscription_id: UUID,
        new_plan_id: UUID
    ) -> bool:
        """Alterar plano de uma assinatura (admin)"""
        subscription = await SubscriptionRepository.find_by_id(session, subscription_id)
        if not subscription:
            return False

        plan = await PlanRepository.find_by_id(session, new_plan_id)
        if not plan or not plan.is_active:
            return False

        subscription.plan_id = new_plan_id
        # Se estiver ativa, atualizamos a expiração baseada no novo plano a partir de hoje
        if subscription.status == "active":
            now = datetime.utcnow()
            subscription.expires_at = now + timedelta(days=30 * plan.duration_months)

        await session.flush()
        await session.commit()
        logger.info(f"Plano da assinatura {subscription_id} alterado para {new_plan_id}")
        return True



class PaymentCronService:
    """Serviço de tarefas agendadas (cron jobs)"""

    @staticmethod
    async def expire_subscriptions(session: AsyncSession) -> int:
        """Rodar diariamente para expirar assinaturas vencidas"""
        count = await SubscriptionRepository.expire_outdated(session)
        if count > 0:
            await session.commit()
            logger.info(f"Assinaturas expiradas: {count}")
        return count

    @staticmethod
    async def finalize_canceled(session: AsyncSession) -> int:
        """Rodar diariamente para finalizar cancelamentos"""
        count = await SubscriptionRepository.finalize_canceled(session)
        if count > 0:
            await session.commit()
            logger.info(f"Assinaturas canceladas (finalizadas): {count}")
        return count
